quickwit-metastore/src/metastore/single_file_metastore/mod.rs

```diff
@@ -0,0 +1,575 @@
+                *metadata_set_lock = metadata_set;
+                Ok(())
+            }
+            err @ Err(_) => {
+                // metadata_set_mutex.discard();
+                // For some of the error type here, we cannot know for sure
+                // whether the content was written or not.
+                //
+                // Just to be sure, let's discard the cache.
+                let mut per_index_metastores_wlock = self.per_index_metastores.write().await;
+
+                // At this point, we hold both locks.
+                per_index_metastores_wlock.remove(index_id);
+                metadata_set_lock.discarded = true;
+
+                err
+            }
+        }
+    }
+
+    async fn read<T, F>(&self, index_id: &str, view: F) -> MetastoreResult<T>
+    where F: FnOnce(&MetadataSet) -> MetastoreResult<T> {
+        let metadata_set_lock = self.metadata_set_locked(index_id).await?;
+        view(&*metadata_set_lock)
+    }
+
+    /// Returns a valid metadataset that is locked.
+    ///
+    /// This function guarantees that the metadataset has not been
+    /// marked as discarded.
+    async fn metadata_set_locked(
+        &self,
+        index_id: &str,
+    ) -> MetastoreResult<OwnedMutexGuard<MetadataSet>> {
+        loop {
+            let metadata_set_mutex = self.metadata_set(index_id).await?;
+            let metadata_set_lock = metadata_set_mutex.lock_owned().await;
+            if !metadata_set_lock.discarded {
+                return Ok(metadata_set_lock);
+            }
+        }
+    }
+
+    /// Returns an IndexView for the given index_id.
+    ///
+    /// If this is the first call during this instance for this
+    /// `index_id`, a fetch to the storage will be initiated
+    /// and might triggered an error.
+    ///
+    /// For a given index_id, only copies of the same index_view are returned.
+    async fn metadata_set(&self, index_id: &str) -> MetastoreResult<Arc<Mutex<MetadataSet>>> {
+        {
+            // Happy path!
+            // If the object is already in our cache then we just return a copy
+            let per_index_metastores = self.per_index_metastores.read().await;
+            if let Some(index_view) = per_index_metastores.get(index_id) {
+                return Ok(index_view.clone());
+            }
+        }
+
+        // At this point we do not hold our mutex, so we need to do a little dance
+        // to make sure we return the same instance.
+        //
+        // If there is an error here, note we do not return right away.
+        // That's because we want to observe the property that after one success
+        // all subsequent calls will succeed.
+        let metadata_set_result = fetch_metadata_set(&*self.storage, index_id).await;
+
+        // Here we retake the lock, still no io ongoing.
+        let mut per_index_metastores_wlock = self.per_index_metastores.write().await;
+        if let Some(metadata_set) = per_index_metastores_wlock.get(index_id) {
```

Comment:
Can we comment that at this point the `metadata_set` might be put by another client?

Reply:
Adding a comment